Add va-breadcrumb USWDS v3 variation

harshil1793 opened the pull request:
## Chromatic
<!-- This `{{head.ref}}` is a placeholder for a CI job - it will be updated automatically -->
https://{{head.ref}}--60f9b557105290003b387cd5.chromatic.com

---
## Configuring this pull request
- [ ] Link to any related issues in the description so they can be cross-referenced.
- [ ] Add the appropriate version patch label (`major`, `minor`, `patch`, or `ignore-for-release`). 
    - See [How to choose a version number](https://github.com/department-of-veterans-affairs/component-library#how-to-choose-a-version-number) for guidance.
    - Use `ignore-for-release` if files haven't been changed in a component library package. (ie. only Storybook files)


## Description
Closes [1716](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/1716)

## Testing done
Tested in Storybook
Chrome/Edge/Mozzila

## Screenshots
![breadcrumb](https://github.com/department-of-veterans-affairs/component-library/assets/22892911/686a7026-f559-4f9b-9286-14bf1590fcbf)


## Acceptance criteria
- [ ] A USWDS v3 variation exists for va-breadcrumb web component

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

Andrew565 reviewed (dismissed):
LGTM. Good catch on changing from 'ul' to 'ol' to better match what's in the USWDS V3 version.

jamigibbs commented:
Does this support all three of the examples shown on the USWDS website? 

1. Default
2. With RDFA metadata
3. Wrapping

![Screenshot 2023-06-05 at 10 20 30 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/22ab9651-a1f4-4278-962f-afbe49a75c41)

jamigibbs commented:
I'm seeing lots of `rem` styles for the slot content. Should we convert all of those to pixel so that it'll be consistent for both v1 and v3? 

![Screenshot 2023-06-05 at 10 25 15 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/082af238-9195-48d1-ab8c-0ecfa9b09527)

![Screenshot 2023-06-05 at 10 26 37 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/3c4cddd3-50c0-45bf-97b0-c9be2292b2c3)

jamigibbs commented:
The USWDS classes in their examples looks different than what is in Storybook:

**USWDS**

![Screenshot 2023-06-05 at 10 36 30 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/eaad4a58-d6ca-4b75-80ef-14583390ea4b)

**VA Storybook**

![Screenshot 2023-06-05 at 10 37 19 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/006d5f4e-fce4-43a3-8697-23878f04e4ad)

jamigibbs commented:
> The USWDS classes in their examples looks different than what is in Storybook:
> 
> **USWDS**
> 
> ![Screenshot 2023-06-05 at 10 36 30 AM](https://user-images.githubusercontent.com/872479/243402269-eaad4a58-d6ca-4b75-80ef-14583390ea4b.png)


After the classes have been updated so that they are using the USWDS classes, make sure to also add any that have `rem` values to the formation stylesheet and import it:

`@import '../../global/formation_overrides';`

harshil1793 commented:
With RDFA metadata, Wrapping, I believe both should be controlled by prop, right?

The USWDS and storybook classes should match after I include forward settings.

jamigibbs commented:
> With RDFA metadata, Wrapping, I believe both should be controlled by prop, right?
> 
> The USWDS and storybook classes should match after I include forward settings.

You will need to [look at the examples](https://designsystem.digital.gov/components/breadcrumb/) and determine how to best implement. Perhaps create more Storybook stories with the right classes passed into the slot or something. 

You are also missing `usa-current`

![Screenshot 2023-06-05 at 10 53 37 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/05dddb88-bc80-41dd-9d76-74d3f94cf205)

harshil1793 commented:
@jamigibbs @Andrew565 I encountered some slot classes that I was unable to override in the formation file. As a solution, I converted all rem units to px in the slot file. Additionally, I introduced a new prop called `wrapping` which allows the user to control the wrapping behavior. This was necessary because I noticed two different variations in the USWDS example."

Andrew565 commented:
Looks like you have a failing test, FYI.

jamigibbs commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -144,16 +153,32 @@ export class VaBreadcrumbs {
+    if (uswds) {
+      return (
+        <Host>
+          <nav aria-label={label} class={`usa-breadcrumb ${wrapClass}`}>
```
All of the classes can be added to `wrapClass` like this:

```
    const wrapClass = classnames({
      'usa-breadcrumb': true,
      'usa-breadcrumb--wrap': wrapping,
    });
```

jamigibbs commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs-slot.css:
```diff
@@ -64,3 +65,22 @@ va-breadcrumbs li > a[aria-current='page'] {
+
+va-breadcrumbs .usa-breadcrumb__list-item:last-child::after {
+  content: '';
+}
```
I don't understand why we are adding these USWDS styles here unless we have a VA need to override something. 

```
va-breadcrumbs .usa-breadcrumb__list-item {
  display: inline-block;
  line-height: 1.5;
  margin: 4px 0;
  padding: 4px 0;
  vertical-align: middle;
}
```

In this part, are we intentionally overriding the icon in order to use Font Awesome?

```
va-breadcrumbs .usa-breadcrumb__list-item::after {
  content: '\203A';
  display: inline-block;
  padding: 0 5.6px;
}
```

The reason I'm pointing this out is that we want to pull styles directly from the USWDS modules so that we get their updates automatically. Any time we add/override USWDS styles in the component itself, we lose that so I want to make sure there is a specific reason for doing it.

harshil1793 replied:
Well, I reason I added it because it shows as a list without icon.

jamigibbs replied:
Is there a Sketch file for this component by chance?

When I look at the USWDS example, there is an icon they provide as part of the breadcrumb module. Sometimes we are using our own Fontawesome icons though so your addition would make sense if that's what the Sketch file shows. But if we are using the USWDS icon, we should make sure that works:

![Screenshot 2023-06-06 at 11 48 02 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/3b4af77d-629f-4faf-9ff9-c569d4eb7552)

jamigibbs commented:
I don't think I see `usa-current` styles applying.

![Screenshot 2023-06-06 at 11 37 06 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/e5f65d32-c81a-43c7-932a-33841e0e5e91)

jamigibbs commented:
Can you mention in the documentation text that this prop is for USWDS? ie. "Available when uswds is true."

![Screenshot 2023-06-06 at 11 43 46 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/e164a757-473d-458a-b9e3-be850faa3414)

Andrew565 commented on packages/storybook/stories/va-breadcrumbs-uswds.stories.jsx:
```diff
@@ -0,0 +1,132 @@
+      <button onClick={e => resetCrumbs()}>Reset Crumbs</button>
+      <br />
+      <p>
+        Note: To rerender the breadcrumbs dynamically, the anchor links must be
```
This note doesn't make sense anymore with using an object prop for the breadcrumbs list.

Andrew565 commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -25,7 +28,22 @@ export class VaBreadcrumbs {
+  @Prop() wrapping?: boolean = false;
+
+  @Prop() breadcrumbList?: string;
+  @State() myInnerArray?: Array<{ label: string; href: string }>;
```
This should be something more descriptive like "breadcrumbsState".

jamigibbs commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -25,7 +28,22 @@ export class VaBreadcrumbs {
+     */
+  @Prop() wrapping?: boolean = false;
+
+  @Prop() breadcrumbList?: string;
```
This prop needs documentation.  
<img width="1052" alt="Screenshot 2023-07-06 at 3 36 31 PM" src="https://github.com/department-of-veterans-affairs/component-library/assets/872479/43fc4f2d-06ef-40e8-99c3-836137230895">

harshil1793 commented on packages/web-components/src/global/main.css:
```diff
@@ -1,5 +1,5 @@
 @import 'variables.css';
-@import '../components/va-breadcrumbs/va-breadcrumbs-slot.css';
+@import '../components/va-breadcrumbs/va-breadcrumbs-slot.scss';
```
@Andrew565 some styles are getting messed by importing it in va-breadcrumbs.scss.

Andrew565 commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -25,7 +28,34 @@ export class VaBreadcrumbs {
+   */
+  @Prop() wrapping?: boolean = false;
+  /**
+   *  It represents a list of breadcrumbs with label, href and send as string. This prop is available when `uswds` is set to `true`.
```
Replace with "Represents a list of breadcrumbs. Use an array of objects with `label` and `href` properties, and then use `JSON.stringify()` to convert to a string. This prop is available when `uswds` is set to `true`."

Andrew565 commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -25,7 +28,34 @@ export class VaBreadcrumbs {
+   * 
+   * This is a method that watches for changes in the 'breadcrumbList' prop. 
+   * When the 'breadcrumbList' prop changes, this method parses the new value (provided as a JSON-formatted string) 
+   * into a JavaScript object and assigns it to the 'myInnerArray' state.
```
"... and assigns it to the `breadcrumbsState` state."

Andrew565 commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -81,11 +115,24 @@ export class VaBreadcrumbs {
+        node.classList.add('usa-current');
+        node.replaceChild(span, anchor);
+        /* eslint-disable-next-line i18next/no-literal-string */
+        node?.setAttribute('aria-current', 'page');
```
This doesn't appear to be working, didn't see this attribute in Chromatic. Maybe try moving it up before the replaceChild call?

harshil1793 replied:
I think it was not needed so I added a conditional attribute to fix it and removed that code.

Andrew565 reviewed (changes requested):
Doing a great job, just a few more things to work out.

jamigibbs commented on packages/storybook/stories/va-breadcrumbs-uswds.stories.jsx:
```diff
@@ -0,0 +1,130 @@
+import { getWebComponentDocs, propStructure, StoryDocs } from './wc-helpers';
+import { VaBreadcrumbs } from '@department-of-veterans-affairs/web-components/react-bindings';
+
+const breadcrumbsDocs = getWebComponentDocs('va-breadcrumbs');
```
Can you add this mapping line to the Storybook file?

`VaBreadcrumbs.displayName = 'VaBreadcrumbs';`

This will display the React binding name in the documentation instead of it saying `forwardRef`.

![Screenshot 2023-07-28 at 11 28 09 AM](https://github.com/department-of-veterans-affairs/component-library/assets/872479/9fcdda52-8f70-44c6-a289-fbc0131076a5)

jamigibbs commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -25,7 +28,35 @@ export class VaBreadcrumbs {
+  /**
+   *  Represents a list of breadcrumbs. Use an array of objects with label and href properties, and then use JSON.stringify() to convert to a string. This prop is available when uswds is set to true.
+   */
+  @Prop() breadcrumbList?: any;
```
In the other `uswds` only prop, you wrote the last sentence with code markdown. Can you do the same thing with this one? Just looking for consistency on Storybook.

"This prop is available when `uswds` is set to `true`."

jamigibbs commented on packages/web-components/src/components/va-breadcrumbs/va-breadcrumbs.tsx:
```diff
@@ -25,7 +28,35 @@ export class VaBreadcrumbs {
+   * When the 'breadcrumbList' prop changes, this method parses the new value (provided as a JSON-formatted string) 
+   * into a JavaScript object and assigns it to the 'breadcrumbsState' state.
+   */
+  @Watch('breadcrumbList')
```
Are `@State()` and `@Watch` only used for `uswds`? If so, can you just make a mention in the comments that this is for uswds? That might help us with maintaining the code in the future.

jamigibbs commented on packages/web-components/src/global/main.css:
```diff
@@ -8,4 +8,4 @@ va-process-list li :is(h2, h3, h4, h5) {
   clear: none;
   padding-top: 0.3rem;
-}
+}
```
Is it possible to add the line break back to this file so that it's not showing a change here? If not, no big deal.

jamigibbs reviewed (approved):
Dev looks good to me. Can you add a link to the Sketch file and also have Lilly or Barb (whoever worked on the design) review as well?

Andrew565 reviewed (approved):
LGTM, as this solution represents the best-case scenario given the constraints we faced in implementing it.

babsdenney reviewed (approved):
The design looks great! Thank you!
